### PR TITLE
Add support for Merkleization

### PR DESCRIPTION
**Pull request by emlautarom1**
Implements #77

### CODE REVIEW

File: src/Paprika.Tests/MerkleTests.cs [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        using var db = PagedDb.NativeMemoryDb(SmallDb);
+
+        using (var batch = db.BeginNextBatch())
+        {
+            batch.Commit(CommitOptions.FlushDataAndRoot);
+        }
+
+        using (var batch = db.BeginReadOnlyBatch())
+        {
+            var rootHash = batch.GetRootHash();
```
Comment (Scooletz) [inline src/Paprika.Tests/MerkleTests.cs:23]: I think we should leave the db agnostic in regards to the calculations. As implemented right now in #88 , there is a clear split between blocks that are `latest` or `safe` and blocks that are `finalized`. Only the latter will be stored in the database in a form that makes it squashed together. If you build up the tests maybe it would be useful to build it on level, where you imperatively state `Set for leaves values'. After all, when the left-fold is applied, all the information will be squashed together.

File: src/Paprika/Tree/Leaf.cs [hunk 2]
```diff
@@ -0,0 +1,65 @@
+using Paprika.Crypto;
+using Paprika.Data;
+using Paprika.RLP;
+
+namespace Paprika.Tree;
+
+public static class Leaf
```
Comment (Scooletz) [inline src/Paprika/Tree/Leaf.cs:7]: No tests for it right now? 👁️

File: src/Paprika/Tree/Leaf.cs [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+    private static RlpStream RlpOfAccount(Account account, Keccak codeHash, Keccak storageRootHash)
+    {
+        var contentLength =
+            Rlp.LengthOf(account.Balance)
+            + Rlp.LengthOf(account.Balance)
+            + Rlp.LengthOfKeccakRlp // CodeHash
+            + Rlp.LengthOfKeccakRlp; // StorageRootHash
+
+        Span<byte> rlpBuffer = new byte[Rlp.LengthOfSequence(contentLength)];
```
Comment (Scooletz) [inline src/Paprika/Tree/Leaf.cs:36]: Easy to optimize away later on 😃

**Review by Scooletz — COMMENTED** [review]
I'm trying my best to merge #88 as soon as possible. Once it's there, it should be easier to discuss the finalized data set and more.

File: src/Paprika.Tests/MerkleKeccakTests.cs [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+using Paprika.Tree;
+
+namespace Paprika.Tests;
+
+public class MerkleKeccakTests
+{
+    [Test]
+    [TestCase(0, "E2533A0A0C4F1DDB72FEB7BFAAD12A83853447DEAAB6F28FA5C443DD2D37C3FB")]
+    [TestCase(1, "DD358AF6B1D8E875FBA0E585710D054F14DD9D06FA3C8C5F2BAF66F413178F82")]
+    [TestCase(2, "A654F039A5F9E9F30C89F21555C92F1CB1E739AF11A9E9B12693DEDC6E76F628")]
```
[the remark this answers is not in the document]
Comment (emlautarom1) [reply src/Paprika.Tests/MerkleKeccakTests.cs:14]: That's quite nice. I really disliked using indexes to deal with this.

File: src/Paprika/Tree/Leaf.cs (hunk 2 above)
Comment (emlautarom1) [reply src/Paprika/Tree/Leaf.cs:7]: We have 100% coverage of this class

File: src/Paprika/Tree/KeccakOrRlp.cs [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    public static KeccakOrRlp WrapRlp(Span<byte> data)
+    {
+        var destination = new byte[32];
+
+        if (data.Length < 32)
+        {
+            destination[0] = (byte)data.Length;
+            data.CopyTo(destination[1..]);
+            return new KeccakOrRlp(Type.Rlp, destination);
+        }
```
Comment (emlautarom1) [inline src/Paprika/Tree/KeccakOrRlp.cs:31]: We have no test case that covers this branch. It seems that for Leaves it is impossible, but it's worth mentioning. I initially made a mistake of doing `return new KeccakOrRlp(Type.Rlp, data);` instead of `return new KeccakOrRlp(Type.Rlp, destination);`

File: src/Paprika.Tests/MerkleTests.cs (hunk 1 above)
Comment (emlautarom1) [reply src/Paprika.Tests/MerkleTests.cs:23]: I'm not sure I quite follow. You mean separating `computing/setting the hashes` from `get the root hash`? Could you rephrase the last two sentences?

Comment (Scooletz) [reply src/Paprika.Tests/MerkleTests.cs:23]: I think we could have the batch a bit more agnostic about the `RootHash`. Yes, in #88 there's the metadata, but maybe we could leave it and just operate on the tree with setting `Set(Key.KeccakOrRlp(NibblePath.Empty), ...` ? We can leave as is for now and discuss further. It does not break the idea that much. Let's pause this discussion till #88 lands.

File: src/Paprika/Tree/KeccakOrRlp.cs (hunk 5 above)
Comment (Scooletz) [reply src/Paprika/Tree/KeccakOrRlp.cs:31]: It could be the case if a path is 1 nibble only and the account is almost empty 👀 , like 0.0000....1 Eth and 0 nonce?

File: src/Paprika/Tree/Leaf.cs [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+        var result = EncodeLeaf(nibblePath, rlpStream.Data);
+
+        return result;
+    }
+
+    private static RlpStream RlpOfAccount(Account account, Keccak codeHash, Keccak storageRootHash)
+    {
+        var contentLength =
+            Rlp.LengthOf(account.Balance)
+            + Rlp.LengthOf(account.Balance)
```
Comment (emlautarom1) [inline src/Paprika/Tree/Leaf.cs:32]: This should be ` Rlp.LengthOf(account.Balance) +  Rlp.LengthOf(account.Nonce)`

Comment (emlautarom1) [reply src/Paprika/Tree/Leaf.cs]: Fixed. Also added test cases to trigger this bug

File: src/Paprika/Merkle/Node.cs [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+public enum NodeType : byte
+{
+    Leaf,
+    Extension,
+    Branch
+}
+
+[StructLayout(LayoutKind.Explicit, Pack = 1, Size = Size)]
+public readonly struct Branch
```
Comment (emlautarom1) [inline src/Paprika/Merkle/Node.cs:15]: A Branch is 35 bytes long if we always want to keep the Keccak (which we initially do):
- 1 byte for `Header`, which contains `Dirty` and `NodeType`
- 2 bytes for `Nibbles`, a Bit Set of nibbles
- 32 bytes for `Keccak` (initially, always present)

File: src/Paprika/Merkle/Node.cs [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+    {
+        var mask = 1;
+        for (byte i = 0; i < 16; i++, mask <<= 1)
+        {
+            if ((_nibbles & mask) != 0)
+            {
+                yield return i;
+            }
+        }
+    }
```
Comment (emlautarom1) [inline src/Paprika/Merkle/Node.cs:44]: Initial approach on enumerating the nibbles from the BitSet

File: src/Paprika/Merkle/Node.cs [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+{
+    // Extension: 1 byte for type + dirty, var-length for path (usually it will be a nibble or a few)
+    private const int Size = 33;
+    private const byte IsDirtyFlag = 0b1000;
+    private const byte NodeTypeFlag = 0b0110;
+
+    [FieldOffset(0)]
+    private readonly byte _header;
+
+    public NibblePath Path => default;
```
Comment (emlautarom1) [inline src/Paprika/Merkle/Node.cs:67]: I'm not quite sure how to store a var-length `NibblePath`

File: src/Paprika/Merkle/Node.cs [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+    // TODO: Should we expose ushort? What about a Span of two bytes?
+    public Branch(ushort nibbles)
+    {
+        _header = (byte)NodeType.Branch << 1 | IsDirtyFlag;
+        _nibbles = nibbles;
+    }
+}
+
+[StructLayout(LayoutKind.Explicit, Pack = 1, Size = Size)]
+public readonly struct Extension
```
Comment (emlautarom1) [inline src/Paprika/Merkle/Node.cs:57]: The size of an `Extension` depends on the size of the `NibblePath`.
- 1 byte for Header, which contains Dirty and NodeType
- 1 byte for the `NibblePath` length (N)
- N bytes for the `NibblePath`

If `M` is the maximum size of a NibblePath, then the maximum size of an `Extension` is M + 2.

Note that there is no Keccak here

File: src/Paprika/Merkle/Node.cs [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+
+    public Extension(NibblePath path)
+    {
+        _header = (byte)NodeType.Extension << 1 | IsDirtyFlag;
+    }
+}
+
+[StructLayout(LayoutKind.Explicit, Pack = 1, Size = Size)]
+// Temporary name due to clash with existing `Leaf`
+public readonly struct LeafNode
```
Comment (emlautarom1) [inline src/Paprika/Merkle/Node.cs:80]: The size of a Leaf depends on the length of the NibblePath.

- 1 byte for Header, which contains Dirty and NodeType
- 32 bytes for Keccak
- 1 byte for the NibblePath length (N)
- N bytes for the NibblePath

If M is the maximum size of a NibblePath, then the maximum size of a Leaf is M + 34.

**Comment by emlautarom1** [comment]
The PR will be split up in multiple separate PRs with smaller scope (see #98 for starters)
